Clojure

nhunzaker opened the pull request:
Woo!

This runs excruciatingly slow when executed through Rake. I'm really not sure why, it's quite fast on its own. In any case, I don't want to merge this until I can sort that out, but here it is!

nhunzaker committed:
[add] Clojure

reagent commented on languages/Clojure/deromanize:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+root="./languages/Clojure/lib/deromanize/target/uberjar"
```
Does this mean that it has to be run from the root directory and can't be run from the `languages/Clojure` directory?  If so, you might want to check out [this way](https://github.com/vigetlabs/otp/blob/d29d87978b24ef67f1d70dd6d0072d4314616c25/languages/Julia/decrypt#L4) of getting the path to the current script.

nhunzaker replied:
Thank you, this was such a bad feel

nhunzaker commented on languages/Clojure/lib/romanize/src/romanize/core.clj:
```diff
@@ -0,0 +1,33 @@
+  (let [[value key] (highest n)
+        next (- n value)]
+    (when (> value 0)
+      (str key (romanize next)))))
```
I feel like there's a fun `reduce` to be had here, I'm just not quite there yet

dce replied:
What happens when value gets to zero?

nhunzaker replied:
In the context of `reduce` ? Not sure. For this, the function would return nil. `(str nil)` maps to an empty string

dce replied:
No, sorry, not in context of `reduce`. Just trying to understand the method in its current form. Got it now.

dce replied:
I'm not sure about `reduce` though. I only think about `reduce` in the context of a collection. For this particular function, where you're iterating until a value reaches zero, recursion feels like the right fit.

nhunzaker replied:
I think, ideally, I should be using `loop` and `recur` here.

nhunzaker replied:
I need to dig deeper into that. Best I can tell, `loop/recur` works sort of like a `let` that can re-evaluate itself

dce replied:
https://www.youtube.com/watch?v=5vq1kXLTtII

dce commented:
Getting a whole bunch of the same error when I run the specs:

```
romanize [clojure] % rake
Building all languages ...
  * Entering directory 'languages/Clojure'
    * Running task 'Clojure:clean'... done.
    * Running task 'Clojure:build'... Retrieving org/clojure/clojure/1.6.0/clojure-1.6.0.pom from central
Retrieving org/sonatype/oss/oss-parent/7/oss-parent-7.pom from central
Retrieving org/clojure/clojure/1.6.0/clojure-1.6.0.jar from central
Compiling deromanize.core
Compiling romanize.core
done.
    * Running tests...Exception in thread "main" java.lang.NullPointerException
    at clojure.lang.Numbers.ops(Numbers.java:961)
    at clojure.lang.Numbers.lt(Numbers.java:219)
    at deromanize.core$digitize.doInvoke(core.clj:22)
    at clojure.lang.RestFn.applyTo(RestFn.java:139)
    at clojure.core$apply.invoke(core.clj:624)
    at deromanize.core$_main.doInvoke(core.clj:29)
    at clojure.lang.RestFn.invoke(RestFn.java:397)
    at clojure.lang.AFn.applyToHelper(AFn.java:152)
    at clojure.lang.RestFn.applyTo(RestFn.java:132)
    at deromanize.core.main(Unknown Source)
Exception in thread "main" java.lang.NullPointerException
    at clojure.lang.Numbers.ops(Numbers.java:961)
    at clojure.lang.Numbers.lt(Numbers.java:219)
    at deromanize.core$digitize.doInvoke(core.clj:22)
    ...
```

nhunzaker commented:
`NullPointerException`? Typical... I'll dig in

dce committed:
Fix bug w/ Clojure implementation

The str/split call was leaving an empty string at the beginning
of the list. A call to 'rest' gets rid of it.

nhunzaker committed:
[update] Use seq and chars for deromanize
